aries_vcx/src/common/credentials/encoding.rs

```diff
@@ -1,99 +1,140 @@
+use std::collections::HashMap;
+
+use serde_json::Value;
+
 use crate::errors::error::{AriesVcxError, AriesVcxErrorKind, VcxResult};
 use crate::utils::openssl::encode;
-use std::collections::HashMap;
 
-pub fn encode_attributes(attributes: &str) -> VcxResult<String> {
-    let mut dictionary = HashMap::new();
-    match serde_json::from_str::<HashMap<String, serde_json::Value>>(attributes) {
-        Ok(mut attributes) => {
-            for (attr, attr_data) in attributes.iter_mut() {
-                let first_attr = match attr_data {
-                    // new style input such as {"address2":"101 Wilson Lane"}
-                    serde_json::Value::String(str_type) => str_type,
-
-                    // old style input such as {"address2":["101 Wilson Lane"]}
-                    serde_json::Value::Array(array_type) => {
-                        let attrib_value: &str = match array_type.get(0).and_then(serde_json::Value::as_str) {
-                            Some(x) => x,
-                            None => {
-                                return Err(AriesVcxError::from_msg(
-                                    AriesVcxErrorKind::InvalidAttributesStructure,
-                                    "Attribute value not found",
-                                ));
-                            }
-                        };
-
-                        warn!("Old attribute format detected. See vcx_issuer_create_credential api for additional information.");
-                        attrib_value
-                    }
-                    _ => {
-                        return Err(AriesVcxError::from_msg(
-                            AriesVcxErrorKind::InvalidJson,
-                            "Invalid Json for Attribute data",
-                        ));
-                    }
-                };
-
-                let encoded = encode(first_attr)?;
-                let attrib_values = json!({
-                    "raw": first_attr,
-                    "encoded": encoded
-                });
-
-                dictionary.insert(attr.to_string(), attrib_values);
+/// `CredentialAttributeData` represents contains all the attributes data
+/// in a credential.
+#[derive(Debug, Clone)]
+pub struct CredentialAttributeData<D> {
+    /// A reference to the attributes data in the credentials.
+    data: D,
+}
+
+/// `EncodedCredentialAttributes` contains the encoded credential attributes data.
+#[derive(Debug, Clone)]
+pub struct EncodedCredentialAttributes {
```

Comment:
Hum, having raw data in an encoded type is a bit confusing IMO. What do you think about this?

```rust
pub CredentialAttribute<State> {
    encoded: Option<String>,
    raw: String,
   _marker: PhantomData<State>,
   
}

pub struct  EncodedAttribute;

pub struct RawAttribute;

impl CredentialAttribute<RawAttribute> {
    fn encode(...) -> ...
}

impl CredentialAttribute<Encoded> {
    fn encoded(...) -> ...
}
```

Reply:
We need the `PhantomData` here to workaround unused generic type error since we don't have a field with type `State` in `CredentialAttribute`, but we still want to say that `CredentialAttribute` has a state.

Reply:
Sorry but I don't quite follow. What state does it have? What kind of programmer error are you trying to prevent in compile time by using the `State` generics arg?

Reply:
This is to prevent an API misused. One cannot get encoded data unless they are in an encoded state.

Reply:
Ah I see I get it now - only 
```
CredentialAttribute<RawAttributeValue>
```
implements `encode` functions, while only 
```
CredentialAttribute<EncodedAttributeValue>
```
implements 
```
pub fn encoded(&self) -> VcxResult<&str>
```
So with `CredentialAttribute<EncodedAttributeValue>` we essentially do get the compile time guarantee that `encoded: Option<String>` is `Some`.
It's smart, I haven't used `PhantomData` much in past. Nice!

However I still believe it would be easier to use if the `Option` could be removed - consider if there's a reasonable way to do it, nevertheless I think we can keep the current phantom data approach as well.